File: scripts/_stdlib_yaml.py
"""Minimal stdlib YAML loader for workbench adapters.

Supports:
  - top-level scalars (string, int, bool, null)
  - lists of strings at any indent
  - lists of flat dicts with inline `- key: value` or indented key: value
  - one-level nested string lists inside list items
  - empty lists ([])
  - folded and literal block scalars
"""

from __future__ import annotations

import re
from pathlib import Path
from typing import Any


def _coerce_scalar(value: str) -> Any:
    if value == "":
        return ""
    lower = value.lower()
    if lower == "true":
        return True
    if lower == "false":
        return False
    if lower in ("null", "~"):
        return None
    try:
        return int(value)
    except ValueError:
        pass
    try:
        return float(value)
    except ValueError:
        pass
    if (value.startswith('"') and value.endswith('"')) or (
        value.startswith("'") and value.endswith("'")
    ):
        return value[1:-1]
    return value


def _inline_block_scalars(text: str) -> str:
    lines = text.splitlines()
    out: list[str] = []
    i = 0
    pattern = re.compile(
        r"^(?P<indent>\s*)(?P<prefix>(?:- )?)(?P<key>[A-Za-z_][A-Za-z0-9_-]*):\s*(?P<marker>[>|])\s*$"
    )
    while i < len(lines):
        line = lines[i]
        match = pattern.match(line)
        if not match:
            out.append(line)
            i += 1
            continue
        key_indent = len(match.group("indent"))
        marker = match.group("marker")
        prefix = match.group("prefix")
        key = match.group("key")
        j = i + 1
        content_lines: list[str] = []
        content_indent: int | None = None
        while j < len(lines):
            current = lines[j]
            if current.strip() == "":
                content_lines.append("")
                j += 1
                continue
            current_indent = len(current) - len(current.lstrip(" "))
            if current_indent <= key_indent:
                break
            if content_indent is None:
                content_indent = current_indent
            content_lines.append(current[content_indent:])
            j += 1
        while content_lines and content_lines[-1] == "":
            content_lines.pop()
        if marker == ">":
            parts: list[str] = []
            buffer: list[str] = []
            for content in content_lines:
                if content == "":
                    if buffer:
                        parts.append(" ".join(buffer))
                        buffer = []
                    parts.append("")
                else:
                    buffer.append(content)
            if buffer:
                parts.append(" ".join(buffer))
            inline = " ".join(part for part in parts if part)
        else:
            inline = "\n".join(content_lines)
        escaped = inline.replace("\\", "\\\\").replace('"', '\\"').replace("\n", "\\n")
        out.append(f'{match.group("indent")}{prefix}{key}: "{escaped}"')
        i = j
    return "\n".join(out)


def load(text: str) -> dict[str, Any]:
    text = _inline_block_scalars(text)
    result: dict[str, Any] = {}
    current_key: str | None = None
    current_list: list[Any] | None = None
    pending_obj: dict[str, Any] | None = None
    pending_nested_list_key: str | None = None

    for raw in text.splitlines():
        stripped = raw.strip()
        if not stripped or stripped.startswith("#"):
            continue
        indent = len(raw) - len(raw.lstrip(" "))
        if pending_obj is not None and pending_nested_list_key is not None and indent >= 6 and (
            stripped.startswith("- ") or stripped == "-"
        ):
            item_body = stripped[2:].strip() if stripped.startswith("- ") else ""
            if item_body:
                pending_obj[pending_nested_list_key].append(_coerce_scalar(item_body))
            continue
        if stripped.startswith("- ") or stripped == "-":
            if current_key is None:
                continue
            if current_list is None:
                current_list = []
                result[current_key] = current_list
            item_body = stripped[2:].strip() if stripped.startswith("- ") else ""
            match = re.match(r"^([A-Za-z_][A-Za-z0-9_-]*):\s+(.+)$", item_body)
            if match and not item_body.startswith(('"', "'")):
                obj = {match.group(1): _coerce_scalar(match.group(2).strip())}
                current_list.append(obj)
                pending_obj = obj
                pending_nested_list_key = None
            elif item_body:
                current_list.append(_coerce_scalar(item_body))
                pending_obj = None
                pending_nested_list_key = None
            else:
                pending_obj = {}
                current_list.append(pending_obj)
                pending_nested_list_key = None
            continue
        if pending_obj is not None and indent >= 2 and ":" in stripped:
            key, _, value = stripped.partition(":")
            if not value.strip():
                pending_obj[key.strip()] = []
                pending_nested_list_key = key.strip()
                continue
            pending_obj[key.strip()] = _coerce_scalar(value.strip())
            pending_nested_list_key = None
            continue
        if ":" in stripped and indent == 0:
            current_list = None
            pending_obj = None
            pending_nested_list_key = None
            key, _, value = stripped.partition(":")
            key = key.strip()
            value = value.strip()
            if "  #" in value:
                value = value[: value.index("  #")].strip()
            if not value:
                current_key = key
                continue
            if value == "[]":
                result[key] = []
                current_key = key
                continue
            result[key] = _coerce_scalar(value)
            current_key = key
    return result


def load_file(path: Path) -> dict[str, Any]:
    return load(path.read_text(encoding="utf-8"))

File: scripts/_stdlib_yaml.mjs
import { readFileSync } from "node:fs";

function coerceScalar(value) {
	if (value === "") {
		return "";
	}
	const lower = value.toLowerCase();
	if (lower === "true") {
		return true;
	}
	if (lower === "false") {
		return false;
	}
	if (lower === "null" || lower === "~") {
		return null;
	}
	if (/^-?\d+$/.test(value)) {
		return Number.parseInt(value, 10);
	}
	if (/^-?(?:\d+\.\d*|\d*\.\d+)$/.test(value)) {
		return Number.parseFloat(value);
	}
	if (
		(value.startsWith("\"") && value.endsWith("\"")) ||
		(value.startsWith("'") && value.endsWith("'"))
	) {
		return value.slice(1, -1);
	}
	return value;
}

function trimTrailingEmptyLines(lines) {
	const trimmed = [...lines];
	while (trimmed.length > 0 && trimmed.at(-1) === "") {
		trimmed.pop();
	}
	return trimmed;
}

function foldBlockScalarLines(contentLines) {
	const parts = [];
	let buffer = [];
	for (const content of contentLines) {
		if (content === "") {
			if (buffer.length > 0) {
				parts.push(buffer.join(" "));
				buffer = [];
			}
			continue;
		}
		buffer.push(content);
	}
	if (buffer.length > 0) {
		parts.push(buffer.join(" "));
	}
	return parts.join(" ");
}

function blockScalarContent(lines, startIndex, keyIndent) {
	const contentLines = [];
	let contentIndent = null;
	let index = startIndex;
	for (; index < lines.length; index += 1) {
		const current = lines[index];
		if (current.trim() === "") {
			contentLines.push("");
			continue;
		}
		const currentIndent = current.length - current.trimStart().length;
		if (currentIndent <= keyIndent) {
			break;
		}
		if (contentIndent === null) {
			contentIndent = currentIndent;
		}
		contentLines.push(current.slice(contentIndent));
	}
	return {
		contentLines: trimTrailingEmptyLines(contentLines),
		nextIndex: index,
	};
}

function blockScalarText(marker, contentLines) {
	return marker === ">" ? foldBlockScalarLines(contentLines) : contentLines.join("\n");
}

function escapeQuotedScalar(text) {
	return text
		.replaceAll("\\", "\\\\")
		.replaceAll("\"", "\\\"")
		.replaceAll("\n", "\\n");
}

function inlineBlockScalars(text) {
	const lines = text.split(/\r?\n/u);
	const out = [];
	const pattern =
		/^(?<indent>\s*)(?<prefix>(?:- )?)(?<key>[A-Za-z_][A-Za-z0-9_-]*):\s*(?<marker>[>|])\s*$/u;

	for (let index = 0; index < lines.length; index += 1) {
		const line = lines[index];
		const match = pattern.exec(line);
		if (!match?.groups) {
			out.push(line);
			continue;
		}
		const { contentLines, nextIndex } = blockScalarContent(lines, index + 1, match.groups.indent.length);
		const escaped = escapeQuotedScalar(blockScalarText(match.groups.marker, contentLines));
		out.push(`${match.groups.indent}${match.groups.prefix}${match.groups.key}: "${escaped}"`);
		index = nextIndex - 1;
	}

	return out.join("\n");
}

function isListLine(stripped) {
	return stripped.startsWith("- ") || stripped === "-";
}

function listItemBody(stripped) {
	return stripped.startsWith("- ") ? stripped.slice(2).trim() : "";
}

function stripInlineComment(stripped) {
	if (!stripped.includes("  #")) {
		return stripped;
	}
	return stripped.slice(0, stripped.indexOf("  #")).trimEnd();
}

function prepareLines(text) {
	return inlineBlockScalars(text)
		.split(/\r?\n/u)
		.map((raw) => {
			const stripped = stripInlineComment(raw.trim());
			return {
				raw,
				stripped,
				indent: raw.length - raw.trimStart().length,
			};
		});
}

function skipIgnorable(lines, index) {
	let current = index;
	while (current < lines.length) {
		const stripped = lines[current].stripped;
		if (!stripped || stripped.startsWith("#")) {
			current += 1;
			continue;
		}
		break;
	}
	return current;
}

function parseScalar(value) {
	if (value === "[]") {
		return [];
	}
	if (value === "{}") {
		return {};
	}
	return coerceScalar(value);
}

function splitKeyValue(stripped) {
	const separator = stripped.indexOf(":");
	if (separator === -1) {
		return null;
	}
	return {
		key: stripped.slice(0, separator).trim(),
		value: stripped.slice(separator + 1).trim(),
	};
}

function parseNode(lines, index, indent) {
	const current = skipIgnorable(lines, index);
	if (current >= lines.length || lines[current].indent < indent) {
		return { value: null, index: current };
	}
	if (isListLine(lines[current].stripped)) {
		return parseArray(lines, current, indent);
	}
	return parseObject(lines, current, indent);
}

function shouldStopObjectParse(line, indent) {
	return line.indent < indent || line.indent !== indent || isListLine(line.stripped);
}

function parseObjectEntry(result, lines, current, indent) {
	const line = lines[current];
	if (shouldStopObjectParse(line, indent)) {
		return { stop: true, current };
	}
	const pair = splitKeyValue(line.stripped);
	if (!pair) {
		return { stop: true, current };
	}
	current += 1;
	if (pair.value) {
		result[pair.key] = parseScalar(pair.value);
		return { stop: false, current };
	}
	const next = skipIgnorable(lines, current);
	if (next >= lines.length || lines[next].indent <= indent) {
		result[pair.key] = null;
		return { stop: false, current: next };
	}
	const parsed = parseNode(lines, next, lines[next].indent);
	result[pair.key] = parsed.value;
	return { stop: false, current: parsed.index };
}

function parseObject(lines, index, indent) {
	const result = {};
	let current = index;
	while (current < lines.length) {
		current = skipIgnorable(lines, current);
		if (current >= lines.length) {
			break;
		}
		const parsedEntry = parseObjectEntry(result, lines, current, indent);
		if (parsedEntry.stop) {
			break;
		}
		current = parsedEntry.current;
	}
	return { value: result, index: current };
}

function shouldStopArrayParse(line, indent) {
	return line.indent < indent || line.indent !== indent || !isListLine(line.stripped);
}

function parseNestedArrayItem(lines, current, indent) {
	const next = skipIgnorable(lines, current);
	if (next >= lines.length || lines[next].indent <= indent) {
		return { value: null, current: next };
	}
	const parsed = parseNode(lines, next, lines[next].indent);
	return { value: parsed.value, current: parsed.index };
}

function parseArrayMapValue(lines, current, indent) {
	const next = skipIgnorable(lines, current);
	if (next >= lines.length || lines[next].indent <= indent) {
		return { value: null, current: next };
	}
	const parsed = parseNode(lines, next, lines[next].indent);
	return { value: parsed.value, current: parsed.index };
}

function mergeFollowingObject(item, lines, current, indent) {
	const next = skipIgnorable(lines, current);
	if (next >= lines.length || lines[next].indent <= indent || isListLine(lines[next].stripped)) {
		return { value: item, current };
	}
	const parsed = parseObject(lines, next, lines[next].indent);
	return {
		value: { ...item, ...parsed.value },
		current: parsed.index,
	};
}

function parseArrayKeyValueItem(pair, lines, current, indent) {
	const item = {};
	if (pair.value) {
		item[pair.key] = parseScalar(pair.value);
		return mergeFollowingObject(item, lines, current, indent);
	}
	const parsedValue = parseArrayMapValue(lines, current, indent);
	item[pair.key] = parsedValue.value;
	return mergeFollowingObject(item, lines, parsedValue.current, indent);
}

function parseArrayItem(body, lines, current, indent) {
	if (!body) {
		return parseNestedArrayItem(lines, current, indent);
	}
	const pair = splitKeyValue(body);
	if (pair && !body.startsWith("\"") && !body.startsWith("'")) {
		return parseArrayKeyValueItem(pair, lines, current, indent);
	}
	return { value: parseScalar(body), current };
}

function parseArray(lines, index, indent) {
	const result = [];
	let current = index;
	while (current < lines.length) {
		current = skipIgnorable(lines, current);
		if (current >= lines.length) {
			break;
		}
		const line = lines[current];
		if (shouldStopArrayParse(line, indent)) {
			break;
		}
		const body = listItemBody(line.stripped);
		current += 1;
		const parsedItem = parseArrayItem(body, lines, current, indent);
		result.push(parsedItem.value);
		current = parsedItem.current;
	}
	return { value: result, index: current };
}

export function load(text) {
	const lines = prepareLines(text);
	const start = skipIgnorable(lines, 0);
	if (start >= lines.length) {
		return {};
	}
	return parseNode(lines, start, lines[start].indent).value ?? {};
}

export function loadFile(path) {
	return load(readFileSync(path, "utf-8"));
}
